tools/copy_bom/src/image.rs

```diff
@@ -214,3 +216,169 @@ impl ImageManagement {
+fn remove_redudant_copydir(image_managements: &mut Vec<ImageManagement>) {
+    // reverse, then can save the first operation
+    image_managements.reverse();
+    let mut all_copydirs = HashSet::new();
+    for image_management in image_managements.iter_mut() {
+        let mut tmp = Vec::new();
+        image_management.dirs_to_copy.reverse();
+        for (src, dest) in image_management.dirs_to_copy.iter() {
+            let src_path = PathBuf::from(src);
+            match src_path.file_name() {
+                None => tmp.push((src.clone(), dest.clone())),
+                Some(filename) => {
+                    let dest_dir = PathBuf::from(dest)
+                        .join(filename)
+                        .to_string_lossy()
+                        .to_string();
+                    if !all_copydirs.contains(&dest_dir) {
+                        all_copydirs.insert(dest_dir);
+                        tmp.push((src.clone(), dest.clone()));
+                    }
+                }
+            }
+        }
+        tmp.reverse();
+        image_management.dirs_to_copy = tmp;
+    }
+    // reverse back
+    image_managements.reverse();
+}
+
+/// If multiple files will be copied to the same destination, only the *LAST* file will be copied.
+/// This function will deal with files to copy and shared objects to copy.
+/// In the same bom file, if shared objects and user-written file has the same destination, the user-written file first.
+fn remove_redundant_copyfile(image_managements: &mut Vec<ImageManagement>) {
+    // reverse the array
+    image_managements.reverse();
+    let mut all_files = HashSet::new();
+    for image_management in image_managements.iter_mut() {
+        let mut tmp_files = Vec::new();
+        image_management.files_to_copy.reverse();
+        // files first
+        for (src, dest) in image_management.files_to_copy.iter() {
+            if !all_files.contains(dest) {
+                all_files.insert(dest.clone());
+                tmp_files.push((src.clone(), dest.clone()));
+            }
+        }
+        tmp_files.reverse();
+        image_management.files_to_copy = tmp_files;
+        // then shared object
+        let mut tmp_shared_objects = Vec::new();
+        image_management.shared_objects_to_copy.reverse();
+        for (src, dest) in image_management.shared_objects_to_copy.iter() {
+            if !all_files.contains(dest) {
+                all_files.insert(dest.clone());
+                tmp_shared_objects.push((src.clone(), dest.clone()));
+            }
+        }
+        tmp_shared_objects.reverse();
+        image_management.shared_objects_to_copy = tmp_shared_objects;
+    }
+    // reverse back
+    image_managements.reverse();
+}
+
+// remove redundant items in a vec. For duplicate items, only the first item will be reserved
+fn remove_redundant_items_in_vec<T>(raw: &Vec<T>, excludes: Iter<'_, T>) -> Vec<T> 
```

Comment:
Why do you need the parameter excludes?

Reply:
This is because items are in different vectors. It an item is in one vector, the same item should not in all other vectors. We use the excludes to indicate items that already exist in other vectors.